xcode/Subconscious/Shared/Parsers/Tape.swift

```diff
@@ -84,10 +84,23 @@ struct Tape {
     /// Peek forward, and consume if substrings match
     mutating func consumeMatch(_ subsequence: Substring) -> Bool {
         if let endIndex = offset(by: subsequence.count) {
-            if rest[currentIndex..<endIndex] == subsequence {
-                self.currentIndex = endIndex
-                return true
-            }
+            var idxA = currentIndex
+            var idxB = subsequence.startIndex
+            
+            // iterating through both indices is MUCH faster than constructing and comparing
+            // a substring slice, especially since we get early exist on the first mismatched
+            // character.
+            repeat {
+                if rest[idxA] != subsequence[idxB] {
+                    return false
+                }
+                
+                idxA = rest.index(after: idxA)
+                idxB = subsequence.index(after: idxB)
+            } while idxB < subsequence.endIndex
+            
+            self.currentIndex = endIndex
+            return true
```

Comment:
I found this one speedup to be equally effective vs. storing an array of indices (a la https://maximeremenko.com/string-random-access)

Reply:
Interesting!

There may still be an advantage to storing indices in an array once at Tape creation *IFF* it has the same performance characteristics.

```swift
this.indices = Array(base.indices)
```

The reason is that `consumeMatch` is just one place where we want to do this kind of operation. I think we use it for most of the parser, but there are other cases where we might want to use `peek` or other helpers.

We also use `Substring` and `Substring.Index` in a bookkeeping role, for example, to keep track of the `rest` position and `startIndex`. I'm wondering if we could get a performance improvement by storing indices in an array, and thus avoiding the manipulation of `Substring` indexes altogether.

- `rest()` could become a `Range<Int>` that indexes into `indices`
- `advance()` coud increment an `Int` counter by one
- Etc

Reply:
Let's leave that for a future exploration.